### PR TITLE
Add support for CGImageCreateWithMask.

### LINKED ISSUE #1425: Implement CGImageCreateWithMask

### PR DESCRIPTION
**Pull request by DHowett-MSFT**
Fixes #1425.

---
This change is [Reviewable](https://reviewable.io/reviews/microsoft/winobjc/1555)

### CODE REVIEW

File: Frameworks/CoreGraphics/CGImage.mm [hunk 1]
```diff
@@ -525,13 +525,33 @@ CGImageRef CGImageCreateCopyWithColorSpace(CGImageRef ref, CGColorSpaceRef color
...
+ @Notes This function does not defer the composition of its mask.
+        This function supports non-grayscale alpha masks, unlike the reference platform.
  */
 CGImageRef CGImageCreateWithMask(CGImageRef image, CGImageRef mask) {
-    // TODO  #1425 - masks are applied during rendering via D2D.
     RETURN_NULL_IF(!image);
-    UNIMPLEMENTED();
-    return StubReturn();
+    RETURN_NULL_IF(!mask);
+    RETURN_NULL_IF(CGImageIsMask(image));
```
Comment (msft-Jeyaram) [inline Frameworks/CoreGraphics/CGImage.mm]: >CGImageIsMask [](start = 19, length = 13)

return null if it's not a mask?

Comment (DHowett-MSFT) [inline Frameworks/CoreGraphics/CGImage.mm]: `image` is the input image, it is contractually disallowed from being a mask.

File: Frameworks/CoreGraphics/CGImage.mm [hunk 2]
```diff
@@ -525,13 +525,33 @@ CGImageRef CGImageCreateCopyWithColorSpace(CGImageRef ref, CGColorSpaceRef color
...
+
+    size_t width = CGImageGetWidth(image);
+    size_t height = CGImageGetHeight(image);
+
+    woc::unique_cf<CGContextRef> context{ CGBitmapContextCreate(nullptr,
```
Comment (DHowett-MSFT) [inline Frameworks/CoreGraphics/CGImage.mm]: good idea. this will return a cascading null, but it's better to be safe.

Comment (msft-Jeyaram) [inline Frameworks/CoreGraphics/CGImage.mm]: Yep, I'm thinking if someone decides to change the bitmap info (without realizing we don't support all of it due to D2D restrictions), we would want to catch that failure fast.

---
In reply to: [92708281](https://github.com/Microsoft/WinObjC/pull/1555#discussion_r92708281) [](ancestors = 92708281)

File: Frameworks/CoreGraphics/CGImage.mm (hunk 1 above)
Comment (msft-Jeyaram) [inline Frameworks/CoreGraphics/CGImage.mm]: Ops, LOL I was thinking it was mask.

Can you add a check to see if the mask is really as mask?

---
In reply to: [92708221](https://github.com/Microsoft/WinObjC/pull/1555#discussion_r92708221) [](ancestors = 92708221)

Comment (msft-Jeyaram) [inline Frameworks/CoreGraphics/CGImage.mm]: some how my comment won't appear:

Ops, LOL I was thinking it was mask.

Can you add a check to see if the mask is really as mask?

Comment (DHowett-MSFT) [inline Frameworks/CoreGraphics/CGImage.mm]: See notes above; the reference platform requires a _mask_ (black/white image). We add support for an alpha clip (alpha image), but an alpha image _is not a mask_.

File: Frameworks/CoreGraphics/CGImage.mm (hunk 2 above)
Comment (DHowett-MSFT) [inline Frameworks/CoreGraphics/CGImage.mm]: They can't change the bitmap info. Do you mean some internal consumer?

**Comment by MSFTFox** [comment]
:shipit:

**Pull request merged**